paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java

```diff
@@ -250,21 +271,64 @@ public void commit(ManifestCommittable committable, Map<String, String> properti
             // This optimization is mainly used to decrease the number of times we read from files.
             if (safeLatestSnapshotId != null) {
                 baseEntries.addAll(appendTableFiles);
-                noConflictsOrFail(latestSnapshot.commitUser(), baseEntries, compactTableFiles);
+                try {
+                    noConflictsOrFail(latestSnapshot.commitUser(), baseEntries, compactTableFiles);
+                } finally {
+                    long commitDuration = (System.nanoTime() - started) / 1_000_000;
+                    reportCommit(
+                            appendTableFiles,
+                            appendChangelog,
+                            Collections.emptyList(),
+                            Collections.emptyList(),
+                            commitDuration,
+                            generatedSnapshot,
+                            attempts + 1);
+                }
```

Comment:
Don't wrap each `noConflictsOrFail` with `try...finally...`, just wrap the whole function with one `try...finally...` and report the metrics in `finally`.